packages/react-resizable-panels/package.json

```diff
@@ -9,11 +9,33 @@
     "url": "git+https://github.com/bvaughn/react-resizable-panels.git"
   },
   "source": "src/index.ts",
-  "main": "dist/react-resizable-panels.js",
-  "module": "dist/react-resizable-panels.module.js",
-  "types": "dist/react-resizable-panels.d.ts",
+  "main": "dist/react-resizable-panels.cjs.js",
+  "module": "dist/react-resizable-panels.esm.js",
+  "exports": {
+    ".": {
+      "types": {
+        "import": "./dist/react-resizable-panels.cjs.mjs",
+        "default": "./dist/react-resizable-panels.cjs.js"
+      },
```

Comment:
I don't think I understand why this "types" key exists or what it means

Reply:
`package.json#exports` define **everything** about the package (well, almost, but let's simplify here) that is. It's like a higher-priority manifest describing the content of the package.

Consider this:
```json
{
  "name": "my-lib",
  "main": "./lib/main.js",
  "exports": "./lib/public.js"
}
```

In here, `package.json#main` is completely ignored. `package.json#exports` win and that's all. This should be the behavior of every tool that understands `exports`. As soon as a package adds that it opts into that new behavior - other keys might only be used as "fallback" for tools from the pre-exports era.

However, TS has chosen a different route - likely because following this strictly would be quite disruptive for the ecosystem. So despite the fact that modern TS versions know what to do with `exports`... they ignore it *by default*. Most projects are configured with `moduleResolution: node` and that ignores `package.json#exports` completely.

But as soon as you opt in to the new semantics with `moduleResolution: node16`, they start matching the "runtime semantics" - that is: if a package has `exports` then other fields are ignored. And yes, that includes the top-level `package.json#types`. 

Note that `types` in exports are not mandatory (just like `package.json#types` isn't). You can alternatively rely on the "sibling lookup" (TS always will try to load `file.d.ts` when importing~ `file.js`, `file.d.mts` when importing `file.mjs`, etc). Personally, I find the explicit `types` field (within exports or the top-level one) to be more explicit and easier to reason about.

Reply:
Just want to add on to @Andarist's explanation, the `types` condition is here specifically so that the types correctly resolve regardless of the conditions specified in https://www.typescriptlang.org/tsconfig#customConditions (currently it would technically work without the `types` condition anyway because of a [TypeScript bug](https://github.com/microsoft/TypeScript/issues/50762) but that shouldn't be relied on)